src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+                                    VarNameIdentifier,
+                                    SyntaxFactory.SingletonSeparatedList(
+                                        SyntaxFactory.VariableDeclarator(
+                                            joinClause.Into.Identifier,
+                                            argumentList: null,
+                                            SyntaxFactory.EqualsValueClause(
+                                                SyntaxFactory.AnonymousObjectCreationExpression(
+                                                    SyntaxFactory.SeparatedList(
+                                                        processingState.VariableNames.Select(vn => CreateAnonymousObjectMemberDeclarator(vn)))))))));
+                        processingState.AddStatementProcessingMethod(s => CreateJoinStatement(joinClause, AddToBlockTop(intoDeclaration, s)));
+                    }
+                    else
+                    {
+                        processingState.AddStatementProcessingMethod(s => CreateJoinStatement(joinClause, s));
+                    }
+
+                    return true;
+                }
+
+                return false;
+            }
+
+            private static BlockSyntax AddToBlockTop(StatementSyntax newStatement, StatementSyntax statement)
+            {
+                if (statement is BlockSyntax block)
+                {
+                    return block.WithStatements(block.Statements.Insert(0, newStatement));
+                }
+                else
+                {
+                    return SyntaxFactory.Block(newStatement, statement);
+                }
+            }
+
+            private static BlockSyntax WrapWithBlock(StatementSyntax statement)
+                => statement is BlockSyntax block ? block : SyntaxFactory.Block(statement);
+
+            private static ExpressionSyntax ProcessCondition(ExpressionSyntax expression)
+            {
+                if (expression.Kind() == SyntaxKind.ParenthesizedExpression)
+                {
+                    expression = ((ParenthesizedExpressionSyntax)expression).Expression;
+                }
+
+                return expression.WithoutTrivia();
+            }
+
+            private static AnonymousObjectMemberDeclaratorSyntax CreateAnonymousObjectMemberDeclarator(string name)
+                => SyntaxFactory.AnonymousObjectMemberDeclarator(
+                    SyntaxFactory.NameEquals(SyntaxFactory.IdentifierName(name)), SyntaxFactory.IdentifierName(name));
+
+            private static StatementSyntax CreateJoinStatement(JoinClauseSyntax joinClause, StatementSyntax statement)
+            {
+                return SyntaxFactory.Block(
+                                SyntaxFactory.ForEachStatement(
+                                    joinClause.Type ?? VarNameIdentifier,
+                                    joinClause.Identifier,
+                                    joinClause.InExpression,
```

Comment:
>joinClause.InExpression [](start = 36, length = 23)

It is wrong to put this expression into a nested ```for each```, it should be evaluated only once

Reply:
Why?

Input:
from num in new int[] { 1, 2 }
                 join x1 in new int[] { 3, 4 } on num equals x1
                 select x1 + 5

Output:
            foreach (var num in new int[] { 1, 2 })
            {
                foreach (var x1 in new int[] { 3, 4 })
                {
                    if (object.Equals(num, x1))
                    {
                        yield return x1 + 5;
                    }
                }
            }

The InExpression is int[] { 3, 4 }. Where should we evaluate it?

---
In reply to: [177205448](https://github.com/dotnet/roslyn/pull/25362#discussion_r177205448) [](ancestors = 177205448)

Reply:
```new int[] { 3, 4 }``` is evaluated once per query execution. After refactoring it is evaluated once per item in ```int[] { 1, 2 }```. If that was an empty array, then never.

---
In reply to: [177577186](https://github.com/dotnet/roslyn/pull/25362#discussion_r177577186) [](ancestors = 177577186,177205448)